share-coordinator/src/main/java/org/apache/kafka/coordinator/share/ShareCoordinatorShard.java

```diff
@@ -537,38 +539,195 @@ private static ShareGroupOffset merge(ShareGroupOffset soFar, ShareUpdateValue n
+            .build();   
     }
 
     /**
-     * Util method which takes in 2 collections containing {@link PersisterOffsetsStateBatch}
+     * Util method which takes in 2 collections containing {@link PersisterStateBatch}
      * and the startOffset.
-     * It removes all batches from the 1st collection which have the same first and last offset
-     * as the batches in 2nd collection. It then creates a final list of batches which contains the
-     * former result and all the batches in the 2nd collection. In set notation (A - B) U B (we prefer batches in B
-     * which have same first and last offset in A).
-     * Finally, it removes any batches where the lastOffset < startOffset, if the startOffset > -1.
-     * @param currentBatch - collection containing current soft state of batches
-     * @param newBatch - collection containing batches in incoming request
+     * This method checks any overlap between current state batches and new state batches.
+     * Based on various conditions it creates new non-overlapping records preferring new batches.
+     * Finally, it removes any batches where the lastOffset < startOffset, if the startOffset > -1 and
+     * merges any contiguous intervals with same state.
+     * @param batchesSoFar - collection containing current soft state of batches
+     * @param newBatches - collection containing batches in incoming request
      * @param startOffset - startOffset to consider when removing old batches.
      * @return List containing combined batches
      */
-    private static List<PersisterOffsetsStateBatch> combineStateBatches(
-        Collection<PersisterOffsetsStateBatch> currentBatch,
-        Collection<PersisterOffsetsStateBatch> newBatch,
+    // visibility for testing
+    static List<PersisterStateBatch> combineStateBatches(
+        List<PersisterStateBatch> batchesSoFar,
+        List<PersisterStateBatch> newBatches,
         long startOffset
     ) {
-        currentBatch.removeAll(newBatch);
-        List<PersisterOffsetsStateBatch> batchesToAdd = new LinkedList<>(currentBatch);
-        batchesToAdd.addAll(newBatch);
-        // Any batches where the last offset is < the current start offset
-        // are now expired. We should remove them from the persister.
+        // will take care of overlapping batches
+        Queue<PersisterStateBatch> batchQueue = new LinkedList<>(
+            mergeBatches(
+                pruneBatches(
+                    getSortedList(batchesSoFar),
+                    startOffset
+                )
+            ));
+
+        // will take care of overlapping batches
+        List<PersisterStateBatch> modifiedNewBatches = new ArrayList<>(
+            mergeBatches(
+                pruneBatches(
+                    getSortedList(newBatches),
+                    startOffset
+                )
+            ));
+
+        for (PersisterStateBatch batch : modifiedNewBatches) {
+            for (int i = 0; i < batchQueue.size(); i++) {
```

Comment:
We will re-evaluate `batchQueue.size()` on each loop iteration, and it could in theory mutate quite significantly. I don't believe this is a safe loop condition.